fluent-langneg/src/matches.js

```diff
@@ -0,0 +1,210 @@
+ *
+ *    What it does ([requested] * [available] = [supported]):
+ *
+ *    ['en-US'] * ['en'] = ['en']
+ *
+ *  2) likely subtags from LDML 4.3 Likely Subtags has been added
+ *
+ *    The most obvious likely subtag that can be computed is a duplication
+ *    of the language field onto region field (`fr` => `fr-FR`).
+ *
+ *    On top of that, likely subtags may use a list of mappings, that
+ *    allow the algorithm to handle non-obvious matches.
+ *    For example, making sure that we match `en` to `en-US` or `sr` to
+ *    `sr-Cyrl`, while `sr-RU` to `sr-Latn-RU`.
+ *
+ *    This list can be taken directly from CLDR Supplemental Data.
+ *
+ *    What it does ([requested] * [available] = [supported]):
+ *
+ *    ['fr'] * ['fr-FR'] = ['fr-FR']
+ *    ['en'] * ['en-US'] = ['en-US']
+ *    ['sr'] * ['sr-Latn', 'sr-Cyrl'] = ['sr-Cyrl']
+ *
+ *  3) variant/region range check has been added
+ *
+ *    Lastly, the last form of check is against the requested locale ID
+ *    but with the variant/region field replaced with a `*` range.
+ *
+ *    The rationale here laid out in LDML 4.4 Language Matching:
+ *      "(...) normally the fall-off between the user's languages is
+ *      substantially greated than regional variants."
+ *
+ *    In other words, if we can't match for the given region, maybe
+ *    we can match for the same language/script but other region, and
+ *    it will in most cases be preferred over falling back on the next
+ *    language.
+ *
+ *    What it does ([requested] * [available] = [supported]):
+ *
+ *    ['en-AU'] * ['en-US'] = ['en-US']
+ *    ['sr-RU'] * ['sr-Latn-RO'] = ['sr-Latn-RO'] // sr-RU -> sr-Latn-RU
+ *
+ *    It works similarly to getParentLocales algo, except that we stop
+ *    after matching against variant/region ranges and don't try to match
+ *    ignoring script ranges. That means that `sr-Cyrl` will never match
+ *    against `sr-Latn`.
+ */
+export default function filterMatches(
+  requestedLocales, availableLocales, strategy, likelySubtags
+) {
+  const supportedLocales = new Set();
+
+  const availableLocalesCache = new Map();
+
+  outer:
+  for (const reqLocStr of requestedLocales) {
+    if (strategy === 'lookup' && supportedLocales.size > 0) {
+      return Array.from(supportedLocales);
+    }
+
+    const reqLocStrLC = reqLocStr.toLowerCase();
+
+    // Attempt to make an exact match
+    // Example: `en-US` === `en-US`
+    for (const availableLocale of availableLocales) {
+      if (reqLocStrLC === availableLocale.toLowerCase()) {
+        supportedLocales.add(availableLocale);
+        if (strategy !== 'matching') {
+          continue outer;
+        }
+      }
+    }
+
+    const requestedLocale = new Locale(reqLocStrLC);
+
+    // Attempt to match against the available range
+    // This turns `en` into `en-*-*-*` and `en-US` into `en-*-US-*`
+    // Example: ['en-US'] * ['en'] = ['en']
+    for (const availableLocale of availableLocales) {
```

Comment:
It took me quite a while to untangle this code, mostly because the patterns in the code aren't factored in.

Is there a way to make this a second inner loop, between this and the inner
```
if (matches) {
  add_and_bail_if_not_matching
}
```
? Might make the code shorter, and easier to digest, I think